flink-contrib/flink-timerserivce-rocksdb/src/main/java/org/apache/flink/contrib/streaming/api/operators/RocksDBInternalTimerService.java

```diff
@@ -0,0 +1,797 @@
+			int startKeyGroup = keyGroupRange.getStartKeyGroup() + (int)(numFatPartitions * (numKeysPerPartition + 1L) + (partitionIndex - numFatPartitions) * numKeysPerPartition);
+			int endKeyGroup = (int)(startKeyGroup + numKeysPerPartition - 1L);
+
+			return (startKeyGroup > endKeyGroup ? null : new KeyGroupRange(startKeyGroup, endKeyGroup));
+		} else {
+			int startKeyGroup = keyGroupRange.getStartKeyGroup() + (int)(partitionIndex * (numKeysPerPartition + 1L));
+			int endKeyGroup = (int)(startKeyGroup + numKeysPerPartition);
+
+			return new KeyGroupRange(startKeyGroup, endKeyGroup);
+		}
+	}
+
+	// ------------------------------------------------------------------------
+	//  Serialization Methods
+	// ------------------------------------------------------------------------
+
+	private byte[] serializeKeyGroup(int keyGroup) {
+		ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+		DataOutputViewStreamWrapper outputView = new DataOutputViewStreamWrapper(outputStream);
+
+		try {
+			IntSerializer.INSTANCE.serialize(keyGroup, outputView);
+		} catch (IOException e) {
+			throw new RuntimeException("Error while deserializing the key group.", e);
+		}
+
+		return outputStream.toByteArray();
+	}
+
+	private byte[] serializeRawTimer(Tuple4<Integer, Long, K, N> rawTimer) {
+		ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+		DataOutputViewStreamWrapper outputView = new DataOutputViewStreamWrapper(outputStream);
+
+		try {
+			IntSerializer.INSTANCE.serialize(rawTimer.f0, outputView);
+			LongSerializer.INSTANCE.serialize(rawTimer.f1, outputView);
+			keySerializer.serialize(rawTimer.f2, outputView);
+			namespaceSerializer.serialize(rawTimer.f3, outputView);
+		} catch (IOException e) {
+			throw new RuntimeException("Error while serializing the raw timer.", e);
+		}
+
+		return outputStream.toByteArray();
+	}
+
+	private Tuple4<Integer, Long, K, N> deserializeRawTimer(byte[] bytes) {
+		ByteArrayInputStream inputStream = new ByteArrayInputStream(bytes);
+		DataInputViewStreamWrapper inputView = new DataInputViewStreamWrapper(inputStream);
+
+		try {
+			int keyGroup = IntSerializer.INSTANCE.deserialize(inputView);
+			long timestamp = LongSerializer.INSTANCE.deserialize(inputView);
+			K key = keySerializer.deserialize(inputView);
+			N namespace = namespaceSerializer.deserialize(inputView);
+
+			return new Tuple4<>(keyGroup, timestamp, key, namespace);
+		} catch (IOException e) {
+			throw new RuntimeException("Error while deserializing the raw timer.", e);
+		}
+	}
+
+	/**
+	 * A timer store backed by RocksDB. 
+	 * 
+	 * The timers are stored in RocksDB in the order of key groups. To allow 
+	 * efficient access, the timers are partitioned and the leader of each 
+	 * partition is stored in an in-memory heap. The top of the heap is 
+	 * exactly the first timer to trigger. The heap is updated whenever the 
+	 * partition's leader is updated.
+	 */
+	private class PersistentTimerHeap {
```

Comment:
Following up from this concept. I think you are implementing your own outer heap over e.g. Java's `PriorityQueue` because it lacks for some update mechanism. But I think you might simply reuse `org.apache.flink.runtime.operators.sort.PartialOrderPriorityQueue`, which has a method `adjustTop()` that should fit your needs.

Maybe the following design is nice: you keep a `PartialOrderPriorityQueue<RocksTimerHeap>`, where `RocksTimerHeap` replaces the Tuple4, and acts like a facade to RocksDB, where the heap methods are translated into RocksDB operations and the top timer is always kept in the object `RocksTimerHeap`, so that the `PartialOrderPriorityQueue` can be maintained cheaply.

Another disadvantage of just using Tuple4 is that the field names f0, f1, ... are not very readable in the code, and in a dedicated class, we could use int and long to represent the primitives, which again can save us a little overheads.